examples-src/mem_alloc_cuda_example.cpp

```diff
@@ -0,0 +1,173 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <mpi.h>
+#include <cuda_runtime.h>
+
+int main(int argc, char *argv[])
+{
+    int cuda_device_aware = 0;
+    int cuda_managed_aware = 0;
+    int len = 0, flag = 0;
+    int *managed_buf = NULL;
+    int *device_buf = NULL, *system_buf = NULL;
+    int nranks = 0;
+    MPI_Info info;
+    MPI_Session session;
+    MPI_Group wgroup;
+    MPI_Comm system_comm;
+    MPI_Comm cuda_managed_comm = MPI_COMM_NULL;
+    MPI_Comm cuda_device_comm = MPI_COMM_NULL;
+
+    // Usage mode: REQUESTED
+    MPI_Info_create(&info);
+    MPI_Info_set(info, "mpi_memory_alloc_kinds",
+                       "system,cuda:device,cuda:managed");
+    MPI_Session_init(info, MPI_ERRORS_ARE_FATAL, &session);
+    MPI_Info_free(&info);
+
+    // Usage mode: PROVIDED
+    MPI_Session_get_info(session, &info);
+    MPI_Info_get_string(info, "mpi_memory_alloc_kinds",
+                        &len, NULL, &flag);
+
+    if (flag) {
+        char *val, *valptr, *kind;
+
+        val = valptr = (char *) malloc(len);
+        if (NULL == val) return 1;
+
+        MPI_Info_get_string(info, "mpi_memory_alloc_kinds",
+                            &len, val, &flag);
+
+        while ((kind = strsep(&val, ",")) != NULL) {
+            if (strcasecmp(kind, "cuda:managed") == 0) {
+                cuda_managed_aware = 1;
+            }
+            else if (strcasecmp(kind, "cuda:device") == 0) {
+                cuda_device_aware = 1;
+            }
+        }
+        free(valptr);
+    }
+
+    MPI_Info_free(&info);
+
+    MPI_Group_from_session_pset(session, "mpi://WORLD" , &wgroup);
+
+    // Create a communicator for operations on system memory
+    // Usage mode: ASSERTED
+    MPI_Info_create(&info);
+    MPI_Info_set(info, "mpi_assert_memory_alloc_kinds", "system");
+    MPI_Comm_create_from_group(wgroup,
+            "org.mpi-side-doc.mem-kind.example.system",
```

Comment:
```suggestion
            "org.mpi-forum.side-doc.mem-alloc-kind.cuda-example.system",
```
We should make an effort to ensure all our string tags are unique.

Also, there should probably be a policy for (at least the first few pieces of) all string tags published in any MPI document that is blessed by the MPI Forum -- I'm thinking "org.mpi-forum" irrespective of which document, then "side-doc.short-name-of-doc", then "example-identifier", then "usage-descriptor"
@schulzm @wesbland

Reply:
The suggested policy looks good to me.